File: openacc/fortran/operators.f90
!******************************************
! operators.f90
! based on min-app code written by Oliver Fuhrer, MeteoSwiss
! modified by Ben Cumming, CSCS
! *****************************************

! Description: Contains simple operators which can be used on 3d-meshes

module operators

use mpi
use stats,     only: flops_diff
use data,      only: discretizationT, x_old, options, bndN, bndE, bndS, bndW, domain

implicit none

contains

!==============================================================================

subroutine diffusion(u, s)
    ! arguments
    real (kind=8), intent(in)  :: u(options%nx, options%ny)
    real (kind=8), intent(out) :: s(options%nx, options%ny)

    ! local variables
    real (kind=8) :: alpha, dxs
    integer :: i, j
    integer :: iend, jend
    integer :: stats(MPI_STATUS_SIZE,8)
    integer :: requests(8)
    integer :: num_requests, err

    dxs   = 1000.*(options%dx ** 2)
    alpha = options%alpha
    iend  = options%nx-1
    jend  = options%ny-1

    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    !$acc parallel present(s, u, x_old, bndE, bndW, bndN, bndS, options)
    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

    ! the interior grid points
    !$acc loop
    do j = 2, jend
        !$acc loop
        do i = 2, iend
            s(i,j) = -(4.+alpha) * u(i,j)           &   ! central point
                        + u(i-1, j) + u(i+1, j)     &   ! east and west
                        + u(i, j-1) + u(i, j+1)     &   ! north and south
                        + alpha*x_old(i,j) &
                        + dxs*u(i,j)*(1.0_8 - u(i,j))
        end do
    end do

    ! the east boundary
    i = options%nx
    !$acc loop
    do j = 2, jend
        s(i,j) = -(4.+alpha) * u(i,j)        &
                    + u(i-1, j) + u(i, j-1) + u(i, j+1) &
                    + alpha*x_old(i,j) + bndE(j) &
                    + dxs*u(i,j)*(1.0_8 - u(i,j))
    end do

    ! the west boundary
    i = 1
    !$acc loop
    do j = 2, jend
        s(i,j) = -(4.+alpha) * u(i,j)         &
                    + u(i+1, j) + u(i, j-1) + u(i, j+1) &
                    + alpha*x_old(i,j) + bndW(j) &
                    + dxs*u(i,j)*(1.0_8 - u(i,j))
    end do

    ! the north boundary (plus NE and NW corners)
    j = options%ny
    i = 1 ! NW corner
    s(i,j) = -(4.+alpha) * u(i,j)           &
                + u(i+1, j) + u(i, j-1)     &
                + alpha*x_old(i,j)          &
                + bndW(j) + bndN(i) &
                + dxs*u(i,j)*(1.0_8 - u(i,j))

    ! north boundary
    !$acc loop
    do i = 2, iend
        s(i,j) = -(4.+alpha) * u(i,j)        &
                    + u(i-1, j) + u(i+1, j) + u(i, j-1) &
                    + alpha*x_old(i,j) + bndN(i) &
                    + dxs*u(i,j)*(1.0_8 - u(i,j))
    end do

    i = options%nx ! NE corner
    s(i,j) = -(4.+alpha) * u(i,j)       &
                + u(i-1, j) + u(i, j-1) &
                + alpha*x_old(i,j)      &
                + bndE(j) + bndN(i) &
                + dxs*u(i,j)*(1.0_8 - u(i,j))

    ! the south boundary
    j = 1
    i = 1 ! SW corner
    s(i,j) = -(4.+alpha) * u(i,j)       &
                + u(i+1, j) + u(i, j+1) &
                + alpha*x_old(i,j)      &
                + bndW(j) + bndS(i) &
                + dxs*u(i,j)*(1.0_8 - u(i,j))

    ! south boundary
    !$acc loop
    do i = 2, iend
        s(i,j) = -(4.+alpha) * u(i,j)           &
                    + u(i-1,j  ) + u(i+1,j  )   &
                                 + u(i  ,j+1)   &
                    + alpha*x_old(i,j)          &
                    + bndS(i) &
                    + dxs*u(i,j)*(1.0_8 - u(i,j))
    end do

    i = options%nx ! SE corner
    s(i,j) = -(4.+alpha) * u(i,j)       &
                + u(i-1,j) + u(i, j+1)  &
                + alpha*x_old(i,j)      &
                + bndE(j) + bndS(i) &
                + dxs*u(i,j)*(1.0_8 - u(i,j))

    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    !$acc end parallel
    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

    ! accumulate the flop counts
    ! 8 ops total per point
    flops_diff =  flops_diff                    &
                    + 12 * (options%nx-2) * (options%ny-2) & ! interior points
                    + 11 * (options%nx-2 + options%ny-2) &   ! NESW boundary points
                    + 11 * 4                                 ! corner points
end

end module operators

File: openacc/c/operators.c
//******************************************
// operators.f90
// based on min-app code written by Oliver Fuhrer, MeteoSwiss
// modified by Ben Cumming, CSCS
// *****************************************

// Description: Contains simple operators which can be used on 3d-meshes

#include "data.h"
#include "operators.h"
#include "stats.h"

#define U(j,i)    up[(i) + (j)*nx]
#define S(j,i)    sp[(i) + (j)*nx]
#define X(j,i) x_old[(i) + (j)*nx]

void diffusion(const double* up, double* sp)
{
    //struct discretization_t* options = options;

    //double (*u)[options.nx] = (double(*)[options.nx])up;
    //double (*s)[options.nx] = (double(*)[options.nx])sp;

    //double (*x_old)[options.nx] = (double(*)[options.nx])x_old;
    //double *bndE = bndE, *bndW = bndW;
    //double *bndN = bndN, *bndS = bndS;

    double dxs   = 1000.*options.dx*options.dx;
    double alpha = options.alpha;
    int    iend  = options.nx - 1;
    int    jend  = options.ny - 1;

    int i,j;

    int nx=options.nx;
    int ny=options.ny;

// TODO:  parallel region: folloing should be present:
//            s, u, x_old, bndE, bndW, bndN, bndS, options

// TODO: add acceleration on ALL loops
    // the interior grid points
    for (j = 1; j < jend; j++)
        for (i = 1; i < iend; i++)
        {
            S(j, i) = -(4. + alpha)*U(j,i)               // central point
                                    + U(j,i-1) + U(j,i+1) // east and west
                                    + U(j-1,i) + U(j+1,i) // north and south

                                    + alpha*X(j,i)
                                    + dxs*U(j,i)*(1.0 - U(j,i));
        }
    // the east boundary
    {
        i = options.nx - 1;
        for (j = 1; j < jend; j++)
        {
            S(j, i) = -(4. + alpha) * U(j,i)
                        + U(j, i - 1) + U(j - 1, i) + U(j + 1, i)
                        + alpha*X(j, i) + bndE[j]
                        + dxs * U(j, i) * (1.0 - U(j, i));
        }
    }
    // the west boundary
    {
        i = 0;
        for (j = 1; j < jend; j++)
        {
            S(j, i) = -(4. + alpha) * U(j, i)
                        + U(j, i + 1) + U(j - 1, i) + U(j + 1, i)

                        + alpha*X(j, i) + bndW[j]
                        + dxs*U(j, i) * (1.0 - U(j, i));
        }
    }
    // the north boundary (plus NE and NW corners)
    {
        j = options.ny - 1;

        {
            i = 0; // NW corner
            S(j, i) = -(4. + alpha) * U(j, i)
                        + U(j, i + 1) + U(j - 1, i)

                        + alpha * X(j, i) + bndW[j] + bndN[i]
                        + dxs * U(j, i) * (1.0 - U(j, i));
        }

        // north boundary
        for (i = 1; i < iend; i++)
        {
            S(j, i) = -(4. + alpha) * U(j, i)
                        + U(j, i - 1) + U(j, i + 1) + U(j - 1, i)
                        + alpha*X(j, i) + bndN[i]
                        + dxs * U(j, i) * (1.0 - U(j, i));
        }

        {
            i = options.nx - 1; // NE corner
            S(j, i) = -(4. + alpha) * U(j, i)
                        + U(j, i - 1) + U(j - 1, i)
                        + alpha * X(j, i) + bndE[j] + bndN[i]
                        + dxs * U(j, i) * (1.0 - U(j, i));
        }
    }
    // the south boundary
    {
        j = 0;
        {
            i = 0; // SW corner
            S(j, i) = -(4. + alpha) * U(j, i)
                        + U(j, i + 1) + U(j + 1, i)
                        + alpha * X(j, i) + bndW[j] + bndS[i]
                        + dxs * U(j, i) * (1.0 - U(j, i));
        }
        // south boundary
        for (i = 1; i < iend; i++)
        {
            S(j, i) = -(4. + alpha) * U(j, i)
                        + U(j, i - 1) + U(j, i + 1) + U(j + 1, i)
                        + alpha * X(j, i) + bndS[i]
                        + dxs * U(j, i) * (1.0 - U(j, i));
        }
        //
        {
            i = options.nx - 1; // SE corner
            S(j, i) = -(4. + alpha) * U(j, i)
                        + U(j, i - 1) + U(j + 1, i)
                        + alpha * X(j, i) + bndE[j] + bndS[i]
                        + dxs * U(j, i) * (1.0 - U(j, i));
        }
    }
    // Accumulate the flop counts
    // 8 ops total per point
    flops_diff +=
        + 12 * (options.nx - 2) * (options.ny - 2) // interior points
        + 11 * (options.nx - 2  +  options.ny - 2) // NESW boundary points
        + 11 * 4;                                  // corner points
}

